### PR TITLE
Updating generative model to work more seamlessly with net + background data

### PR DESCRIPTION
**Pull request by dnguyen1196**
# Unlabelled data set
Added the zinc and moses data set, together with their tiny and small versions as `pinot.data.zinc_tiny`, `pinot.data.moses_tiny`. Also modified `pinot.data.utils.batch` to handle both the unlabeled data (zinc, moses) and labeled data (esol).

# CLI tool
Added more functionalities to `pinot.app.cli.py` which can now either load pretrained generative model or to train a new generative model using background data. This also means that generative model now works with `Net`. One can initialize `Net` with a trained or newly initialized `GCNModelVAE`.

`pinot.app.cli.py` can be used as a command line tool as follows:

```
usage: cli.py [-h] [--pretrained_gen_model PRETRAINED_GEN_MODEL]
              [--layer LAYER]
              [--hidden_dims_gvae HIDDEN_DIMS_GVAE [HIDDEN_DIMS_GVAE ...]]
              [--embedding_dim EMBEDDING_DIM]
              [--background_data BACKGROUND_DATA]
              [--n_epochs_generative N_EPOCHS_GENERATIVE]
              [--optimizer_generative OPTIMIZER_GENERATIVE]
              [--lr_generative LR_GENERATIVE] [--save_model SAVE_MODEL]
              [--free_gradient] [--noise_model NOISE_MODEL]
              [--optimizer OPTIMIZER] [--out OUT] [--data DATA]
              [--batch_size BATCH_SIZE] [--lr LR] [--partition PARTITION]
              [--n_epochs N_EPOCHS]

With pretrained generative model::
  --pretrained_gen_model PRETRAINED_GEN_MODEL
                        Parameter file of pretrained generative model

With no pretrained generative model:
  --layer LAYER         Type of graph convolution layer
  --hidden_dims_gvae HIDDEN_DIMS_GVAE [HIDDEN_DIMS_GVAE ...]
                        Hidden dimensions of the convolution layers
  --embedding_dim EMBEDDING_DIM
                        Embedding dimension (dimension of the encoder's
                        output)
  --background_data BACKGROUND_DATA
                        Background data to pre-train generative model on
  --n_epochs_generative N_EPOCHS_GENERATIVE
                        Number of epochs of generative model pre-training
  --optimizer_generative OPTIMIZER_GENERATIVE
                        Optimizer for generative model pre-training
  --lr_generative LR_GENERATIVE
                        Learning rate for generative model pre-training
  --save_model SAVE_MODEL
                        File to save generative model to

Net settings:
  --free_gradient       Allow for updating gradients of pretrained generative
                        model
  --noise_model NOISE_MODEL
                        Noise model for predictive distribution
  --optimizer OPTIMIZER
                        Choice of ptimizer
  --out OUT             Folder to print out results to
  --data DATA           Data set name
  --batch_size BATCH_SIZE
                        Batch size
  --lr LR               Learning rate
  --partition PARTITION
                        Training-testing split
  --n_epochs N_EPOCHS   Number of epochs
```

### CODE REVIEW

File: pinot/app/cli.py [hunk 1]
```diff
@@ -75,20 +120,33 @@ def run(args):
...
-    parser.add_argument('--partition', default='4:1', type=str)
-    parser.add_argument('--n_epochs', default=5)
+    # With pretrained generative model
+    pretrained_gen_group = parser.add_argument_group("With pretrained generative model:")
+    pretrained_gen_group.add_argument('--pretrained_gen_model', default=None, type=str, help="Parameter file of pretrained generative model")
+
+    # With no pretrained generative model
+    group = parser.add_argument_group("With no pretrained generative model")
+    group.add_argument('--layer', type=str, default='GraphConv', help="Type of graph convolution layer")
+    group.add_argument('--hidden_dims_gvae',nargs='+', type=int, default=[128, 128], help="Hidden dimensions of the convolution layers")
```
Comment (yuanqing-wang) [inline pinot/app/cli.py:130]: shall we keep the door open for sneak in activations and dropout for this? in the same fashion as Net

File: pinot/generative/torch_gvae/decoder.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        neural networks that map latent encoding of node $z_x$ to node
+        predictions $\hat{x}$
+    """
+
+    def __init__(self, embedding_dim, num_atom_types):
+        super(SequentialDecoder, self).__init__()
+        self.embedding_dim = embedding_dim
+        self.num_atom_types = num_atom_types
+
+        self.Dx1 = 64
```
Comment (yuanqing-wang) [inline pinot/generative/torch_gvae/decoder.py:74]: let's not hard code anything here. I think it's good enough to leave `z_to_zx` and friends as modules.

or, we can have them to be either torch models or a list of config string in the style of `pinot.representation`

File: pinot/app/cli.py [hunk 3]
```diff
@@ -6,17 +6,66 @@
...
+        net_representation = GCNModelVAE(input_feat_dim, 
+                gcn_type=args.layer,
+                gcn_hidden_dims=args.hidden_dims_gvae,
+                embedding_dim=args.embedding_dim)
+
+        # And then train this model
+        gen_optimizer = pinot.app.utils.optimizer_translation(
+            args.optimizer_generative,
+            lr=args.lr_generative)(net_representation)
+        print("Doing training...")
```
[the remark this answers is not in the document]
Comment (dnguyen1196) [reply pinot/app/cli.py]: yeah we can do that, will move to logging

File: pinot/app/cli.py (hunk 1 above)
Comment (dnguyen1196) [reply pinot/app/cli.py:130]: yeah, we certainly can

File: pinot/generative/torch_gvae/decoder.py (hunk 2 above)
Comment (dnguyen1196) [reply pinot/generative/torch_gvae/decoder.py:74]: yup, will move these into constructor variables
